Sources/COWMacros/COWExpansionFactory.swift

```diff
@@ -218,30 +223,196 @@ internal class COWExpansionFactory<Context: MacroExpansionContext> {
+    }) else {
+      return
     }
     
+    // Bail out if `appliedStructDecl` does not manually implement `==`.
+    guard let nestingEqualFunc = appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap({$0.as(FunctionDeclSyntax.self)})
+      .first(where: {$0.likelyToConformToEquatable(for: appliedStructDecl)}) else {
+      return
+    }
+    
+    // Another compiler bug another workaround.
+    // The compiler doesn't call the manually-implemented `==` in `appliedStructDecl`;
+    // it will synthesize one for the derived storage and call it instead.
+    // Expand an `==` that is an exact copy (except for the signature) of
+    // `appliedStructDecl` to make sure the program behavior is correct.
+    
+    // In order to make this workaround work with computed properties and
+    // immutable functions with return values, we also have to copy them into
+    // the derived storage.
+    appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap { $0.as(VariableDeclSyntax.self) }
+      .filter { !$0.isStatic && !$0.isStored }
+      .forEach { members.append(MemberBlockItemSyntax(decl: $0)) }
+    appliedStructDecl.memberBlock.members
+      .lazy
+      .map(\.decl)
+      .compactMap { $0.as(FunctionDeclSyntax.self) }
+      .filter { !$0.isStatic && !$0.isMutating && !$0.returnTypeEquals(to: "Void") }
+      .forEach { members.append(MemberBlockItemSyntax(decl: $0)) }
+    
+    guard let body = nestingEqualFunc.as(FunctionDeclSyntax.self)!.body else {
+      // Could this ever happen (i.e. a function declaration without
+      // implementation in struct is invalid)?
+      fatalError()
+    }
+    // Use signature from proxy protocol if needed. See the comments below
+    // (why we are adding a proxy protocol) for the reason.
+    #if swift(<5.9.2)
+    let nestedEqualFuncSignature: SyntaxNodeString = "static func equalsWorkaround(lhs: Self, rhs: Self) -> Bool"
+    #else
+    let nestedEqualFuncSignature: SyntaxNodeString = "static func == (lhs: Self, rhs: Self) -> Bool"
+    #endif
+    let nestedEqualFunc = try! FunctionDeclSyntax(nestedEqualFuncSignature) {
+      body.statements
+    }
+    members.append(MemberBlockItemSyntax(decl: nestedEqualFunc))
+    
+    // Yet another compiler bug and another workaround :)
+    // https://github.com/apple/swift/issues/66348
+    // The compiler does not recognize expanded `==` as a conformance which is
+    // fixed in Swift 5.9.2.
+    // For clients still using older compilers, work around by conforming a
+    // proxy protocol. Idea credits to
+    // https://github.com/JosephDuffy/HashableMacro/commit/250787664a63ceff83c1c9b5e30e574ada568f2f.
+    #if swift(<5.9.2)
+    let equalsProxyProtocol: TypeSyntax = "COW.COWStorageEquatableWorkaround"
+    protocols.append(InheritedTypeSyntax(type: equalsProxyProtocol))
+    #endif
+  }
+  
+  private func forwardCodingKeysForDerivedStorage(
```

Comment:
I think `forwardCodingKeysForDerivedStorageIfNeeded` is a better name: the suffix `IfNeeded` indicates this function is not the main point of program logic.